File: project_euler/problem_042/solution.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_triangular(int v) {
    // v is triangular iff 8v+1 is a perfect square (Theorem 1).
    if (v <= 0) return v == 0;
    long long disc = 1 + 8LL * v;
    long long s = (long long)sqrt((double)disc);
    while (s * s < disc) s++;
    while (s * s > disc) s--;
    return s * s == disc;
}

int word_value(const string& w) {
    int val = 0;
    for (char c : w) {
        if (c >= 'A' && c <= 'Z') val += c - 'A' + 1;
    }
    return val;
}

int main() {
    vector<string> candidates = {
        "words.txt",
        "0042_words.txt",
        "project_euler/problem_042/words.txt",
        "project_euler/problem_042/0042_words.txt"
    };

    ifstream fin;
    for (const auto& path : candidates) {
        fin.open(path);
        if (fin.is_open()) {
            break;
        }
        fin.clear();
    }

    if (!fin.is_open()) {
        cerr << "Could not locate the Problem 42 word list." << endl;
        return 1;
    }

    string line;
    getline(fin, line);
    fin.close();

    int count = 0;
    stringstream ss(line);
    string token;
    while (getline(ss, token, ',')) {
        string word;
        for (char c : token) {
            if (c != '"') word += c;
        }
        if (is_triangular(word_value(word))) {
            count++;
        }
    }

    cout << count << endl;
    return 0;
}

File: project_euler/problem_042/solution.py
import math
from pathlib import Path


def is_triangular(v):
    """Check if v is a triangular number using the discriminant criterion:
    v = n(n+1)/2 iff 8v+1 is a perfect square.
    """
    if v <= 0:
        return v == 0
    disc = 1 + 8 * v
    s = math.isqrt(disc)
    return s * s == disc


def word_value(word):
    """Compute the alphabetical value of a word: A=1, B=2, ..., Z=26."""
    return sum(ord(c) - ord('A') + 1 for c in word.upper())


def load_words():
    script_dir = Path(__file__).resolve().parent
    candidates = [
        script_dir / "words.txt",
        script_dir / "0042_words.txt",
        Path("words.txt"),
        Path("0042_words.txt"),
        Path("project_euler/problem_042/words.txt"),
        Path("project_euler/problem_042/0042_words.txt"),
    ]

    for path in candidates:
        if path.exists():
            return path.read_text().strip()

    raise FileNotFoundError("Could not locate the Problem 42 word list.")


def solve():
    content = load_words()
    words = [w.strip('"') for w in content.split(',')]
    return sum(1 for w in words if is_triangular(word_value(w)))


answer = solve()
assert answer == 162
print(answer)
